File: Envs.cpp
#include "Envs.h"
Envs::Res Envs::reset() {
    Res res;
    for (int i = 0; i < NUM_ENVS; i++) {
        _envs[i].reset();
//        for(auto e: _envs.at(i).get_the_1st_observation()){
//            res.feature.at(i).emplace_back(static_cast<float>(e));
//        }
        res.feature.at(i) = _envs.at(i).get_the_1st_observation();
        res.reward.emplace_back(0);
        res.done.emplace_back(false);
    }
    return res;
}
vector<vector<double>> Envs::reset(int env_idx) {
     _envs[env_idx].reset();
    return _envs[env_idx].get_the_1st_observation();
}
Envs::Res Envs::step(const std::vector<std::vector<double>> &actions) {
    Res res;
    Drcu::Res drcu_res;
    for (int i = 0; i < NUM_ENVS; i++) {
        drcu_res = _envs[i].step(actions.at(i));
        res.feature.at(i) = drcu_res.feature;
        res.reward.emplace_back(drcu_res.reward);
        res.done.emplace_back(drcu_res.done);
    }
    return res;
}
Envs::Res Envs::init(int argc, char **short_format_argv) {
    Res res;
    std::vector<std::string> argv (argc); 
    for (int i = 0; i < argc; ++i) {
        argv[i] = short_format_argv[i]; 
    }
    for (int i = 0; i < NUM_ENVS; i++) {
        _envs[i].init(argv);
//        for(auto e: _envs.at(i).get_the_1st_observation()){
//            res.feature.at(i).emplace_back(static_cast<float>(e));
//        }
        res.feature.at(i) = _envs.at(i).get_the_1st_observation();
        res.reward.emplace_back(0);
        res.done.emplace_back(false);
    }
    return res;
}

std::array<double, 4> Envs::get_all_vio() const {
    return _envs[0].get_all_vio();
}

File: Envs.py
##
# @file   Envs.py
# @author Yibo Lin
# @date   Feb 2020
#

import numpy as np 
import gym 
from gym import spaces 
import drcu_cpp 

class DrcuEnv (gym.Env):
    def __init__(self, args):
        super(DrcuEnv, self).__init__()
        self.drcu = drcu_cpp.Drcu()
        self.drcu.init(args)
        # feature shape is #nets x #features 
        feature = np.array(self.drcu.get_the_1st_observation())
        #reward = 0 
        #done = False 

        self.action_space = spaces.Box(low=0, high=1, shape=(1,), dtype=np.float32)
        self.observation_space = spaces.Box(low=0, high=1, shape=feature.shape, dtype=np.float32)

    def step(self, action):
        res = self.drcu.step(action)
        return {'feature' : res.feature, 'reward' : res.reward, 'done' : res.done}

    def reset(self):
        self.drcu.reset()

        feature = np.array(self.drcu.get_the_1st_observation())
        reward = 0 
        done = False 
        return {'feature' : feature, 'reward' : reward, 'done' : done}

    def render(self, observation=None):
        vio = self.drcu.get_all_vio() 
        content = "vio: %g, %g, %g, %g" % (vio[0], vio[1], vio[2], vio[3])
        if observation: 
            content += " ; " + str(observation)
        print(content)
